PiTR fail, error: the default cf kv is lost when there is its write cf kv

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
1. do PiTR on cluster B , which use cluster A‘s snapshot and log
2. after cluster B PiTR table filter success, do PiTR on cluster C which use cluster B's snapshot and log

### 2. What did you expect to see? (Required)
PiTR on cluster C failed, error msg like: 
`[ERROR] [restore.go:76] ["failed to restore"] [error="cannot restore the table(Id=2002902) because it is log restored(at 458679524451156942) before snapshot backup(at 458606073366971518). Please respecify the filter that does not contain the table or replace with a newer snapshot backup."]`

### 3. What did you see instead (Required)
PiTR on cluster C failed, error msg:
`[ERROR] [restore.go:76] ["failed to restore"] [error="the default cf kv is lost when there is its write cf kv(db id:2000512, table id:2000461, value UMCA0OfV5eGuBg==)"]`

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

Release Version: v8.5.0-20250609-a850b6f
Edition: Community
Git Commit Hash: a850b6f2634ee51c32b2ab8623c78035186e1278
Git Branch: heads/refs/tags/v8.5.0-20250609-a850b6f
UTC Build Time: 2025-06-09 02:24:01
GoVersion: go1.23.3
Race Enabled: false
Check Table Before Drop: false
Store: tikv

--- a/br/pkg/restore/misc.go
+++ b/br/pkg/restore/misc.go
@@ -62,8 +62,12 @@ type LogRestoreTableIDsBlocklistFile struct {
 	// SnapshotBackupTs records the BackupTS of the PITR restore. Any PITR restore from the log backup of the cluster, whose restoredTS
 	// is less than it, can ignore the restore table IDs blocklist recorded in the file.
 	SnapshotBackupTs uint64 `protobuf:"varint,2,opt,name=snapshot_backup_ts,proto3"`
-	// TableIDs records the table IDs blocklist of the cluster running the log backup task.
+	// RewriteTs records the rewritten timestamp of the meta kvs in this PITR restore.
+	RewriteTs uint64 `protobuf:"varint,6,opt,name=rewrite_ts,proto3"`
+	// TableIds records the table IDs blocklist of the cluster running the log backup task.
 	TableIds []int64 `protobuf:"varint,3,rep,packed,name=table_ids,proto3"`
+	// DbIds records the database IDs blocklist of the cluster running the log backup task.
+	DbIds []int64 `protobuf:"varint,5,rep,packed,name=db_ids,proto3"`
 	// Checksum records the checksum of other fields.
 	Checksum []byte `protobuf:"bytes,4,opt,name=checksum,proto3"`
 }
@@ -106,8 +110,12 @@ func (m *LogRestoreTableIDsBlocklistFile) checksumLogRestoreTableIDsBlocklistFil
 	hasher := sha256.New()
 	hasher.Write(binary.LittleEndian.AppendUint64(nil, m.RestoreCommitTs))
 	hasher.Write(binary.LittleEndian.AppendUint64(nil, m.SnapshotBackupTs))
+	hasher.Write(binary.LittleEndian.AppendUint64(nil, m.RewriteTs))
 	for _, tableId := range m.TableIds {
 		hasher.Write(binary.LittleEndian.AppendUint64(nil, uint64(tableId)))
+	}
+	for _, dbId := range m.DbIds {
+		hasher.Write(binary.LittleEndian.AppendUint64(nil, uint64(dbId)))
 	}
 	return hasher.Sum(nil)
 }
@@ -117,11 +125,13 @@ func (m *LogRestoreTableIDsBlocklistFile) setChecksumLogRestoreTableIDsBlocklist
 }
 
 // MarshalLogRestoreTableIDsBlocklistFile generates an Blocklist file and marshals it. It returns its filename and the marshaled data.
-func MarshalLogRestoreTableIDsBlocklistFile(restoreCommitTs, snapshotBackupTs uint64, tableIds []int64) (string, []byte, error) {
+func MarshalLogRestoreTableIDsBlocklistFile(restoreCommitTs, snapshotBackupTs, rewriteTs uint64, tableIds, dbIds []int64) (string, []byte, error) {
 	blocklistFile := &LogRestoreTableIDsBlocklistFile{
 		RestoreCommitTs:  restoreCommitTs,
 		SnapshotBackupTs: snapshotBackupTs,
+		RewriteTs:        rewriteTs,
 		TableIds:         tableIds,
+		DbIds:            dbIds,
 	}
 	blocklistFile.setChecksumLogRestoreTableIDsBlocklistFile()
 	filename := blocklistFile.filename()
@@ -133,26 +143,26 @@ func MarshalLogRestoreTableIDsBlocklistFile(restoreCommitTs, snapshotBackupTs ui
 }
 
 // unmarshalLogRestoreTableIDsBlocklistFile unmarshals the given blocklist file.
-func unmarshalLogRestoreTableIDsBlocklistFile(data []byte) (restoreCommitTs, snapshotBackupTs uint64, tableIds []int64, err error) {
+func unmarshalLogRestoreTableIDsBlocklistFile(data []byte) (*LogRestoreTableIDsBlocklistFile, error) {
 	blocklistFile := &LogRestoreTableIDsBlocklistFile{}
-	if err = proto.Unmarshal(data, blocklistFile); err != nil {
-		return 0, 0, nil, errors.Trace(err)
+	if err := proto.Unmarshal(data, blocklistFile); err != nil {
+		return nil, errors.Trace(err)
 	}
 	if !bytes.Equal(blocklistFile.checksumLogRestoreTableIDsBlocklistFile(), blocklistFile.Checksum) {
-		return 0, 0, nil, errors.Errorf(
+		return nil, errors.Errorf(
 			"checksum mismatch (calculated checksum is %s but the recorded checksum is %s), the log restore table IDs blocklist file may be corrupted",
 			base64.StdEncoding.EncodeToString(blocklistFile.checksumLogRestoreTableIDsBlocklistFile()),
 			base64.StdEncoding.EncodeToString(blocklistFile.Checksum),
 		)
 	}
-	return blocklistFile.RestoreCommitTs, blocklistFile.SnapshotBackupTs, blocklistFile.TableIds, nil
+	return blocklistFile, nil
 }
 
 func fastWalkLogRestoreTableIDsBlocklistFile(
 	ctx context.Context,
 	s storage.ExternalStorage,
 	filterOutFn func(restoreCommitTs, snapshotBackupTs uint64) bool,
-	executionFn func(ctx context.Context, filename string, restoreCommitTs uint64, tableIds []int64) error,
+	executionFn func(ctx context.Context, filename string, restoreCommitTs, rewriteTs uint64, tableIds, dbIds []int64) error,
 ) error {
 	filenames := make([]string, 0)
 	if err := s.WalkDir(ctx, &storage.WalkOption{SubDir: logRestoreTableIDBlocklistFilePrefix}, func(path string, _ int64) error {
@@ -178,14 +188,14 @@ func fastWalkLogRestoreTableIDsBlocklistFile(
 			if err != nil {
 				return errors.Trace(err)
 			}
-			restoreCommitTs, snapshotBackupTs, tableIds, err := unmarshalLogRestoreTableIDsBlocklistFile(data)
+			blocklistFile, err := unmarshalLogRestoreTableIDsBlocklistFile(data)
 			if err != nil {
 				return errors.Trace(err)
 			}
-			if filterOutFn(restoreCommitTs, snapshotBackupTs) {
+			if filterOutFn(blocklistFile.RestoreCommitTs, blocklistFile.SnapshotBackupTs) {
 				return nil
 			}
-			err = executionFn(ectx, filename, restoreCommitTs, tableIds)
+			err = executionFn(ectx, filename, blocklistFile.RestoreCommitTs, blocklistFile.RewriteTs, blocklistFile.TableIds, blocklistFile.DbIds)
 			return errors.Trace(err)
 		})
 	}
@@ -198,28 +208,40 @@ func CheckTableTrackerContainsTableIDsFromBlocklistFiles(
 	s storage.ExternalStorage,
 	tracker *utils.PiTRIdTracker,
 	startTs, restoredTs uint64,
-	tableNameByTableID func(tableID int64) string,
-	checkTableIDLost func(tableId int64) bool,
+	tableNameByTableId func(tableId int64) string,
+	dbNameByDbId func(dbId int64) string,
+	checkTableIdLost func(tableId int64) bool,
+	checkDBIdlost func(dbId int64) bool,
+	cleanError func(rewriteTs uint64),
 ) error {
 	err := fastWalkLogRestoreTableIDsBlocklistFile(ctx, s, func(restoreCommitTs, snapshotBackupTs uint64) bool {
 		return startTs >= restoreCommitTs || restoredTs <= snapshotBackupTs
-	}, func(_ context.Context, _ string, restoreCommitTs uint64, tableIds []int64) error {
+	}, func(_ context.Context, _ string, restoreCommitTs, rewriteTs uint64, tableIds, dbIds []int64) error {
 		for _, tableId := range tableIds {
 			if tracker.ContainsTableId(tableId) || tracker.ContainsPartitionId(tableId) {
 				return errors.Errorf(
 					"cannot restore the table(Id=%d, name=%s at %d) because it is log restored(at %d) before snapshot backup(at %d). "+
 						"Please respecify the filter that does not contain the table or replace with a newer snapshot backup.",
-					tableId, tableNameByTableID(tableId), restoredTs, restoreCommitTs, startTs)
+					tableId, tableNameByTableId(tableId), restoredTs, restoreCommitTs, startTs)
 			}
 			// the meta kv may not be backed by log restore
-			if checkTableIDLost(tableId) {
-				return errors.Errorf(
-					"cannot restore the table(Id=%d) because it is log restored(at %d) before snapshot backup(at %d). "+
-						"Please respecify the filter that does not contain the table or replace with a newer snapshot backup.",
-					tableId, restoreCommitTs, startTs,
-				)
+			if checkTableIdLost(tableId) {
+				log.Warn("the table is lost in the log backup storage, so that it can not be restored.", zap.Int64("table id", tableId))
 			}
 		}
+		for _, dbId := range dbIds {
+			if tracker.ContainsDB(dbId) {
+				return errors.Errorf(
+					"cannot restore the database(Id=%d, name %s at %d) because it is log restored(at %d) before snapshot backup(at %d). "+
+						"Please respecify the filter that does not contain the database or replace with a newer snapshot backup.",
+					dbId, dbNameByDbId(dbId), restoredTs, restoreCommitTs, startTs)
+			}
+			// the meta kv may not be backed by log restore
+			if checkDBIdlost(dbId) {
+				log.Warn("the database is lost in the log backup storage, so that it can not be restored.", zap.Int64("database id", dbId))
+			}
+		}
+		cleanError(rewriteTs)
 		return nil
 	})
 	return errors.Trace(err)
@@ -233,7 +255,7 @@ func TruncateLogRestoreTableIDsBlocklistFiles(
 ) error {
 	err := fastWalkLogRestoreTableIDsBlocklistFile(ctx, s, func(restoreCommitTs, snapshotBackupTs uint64) bool {
 		return untilTs < restoreCommitTs
-	}, func(ctx context.Context, filename string, _ uint64, _ []int64) error {
+	}, func(ctx context.Context, filename string, _, _ uint64, _, _ []int64) error {
 		return s.DeleteFile(ctx, filename)
 	})
 	return errors.Trace(err)

--- a/br/pkg/restore/misc_test.go
+++ b/br/pkg/restore/misc_test.go
@@ -157,7 +157,7 @@ func TestLogRestoreTableIDsBlocklistFile(t *testing.T) {
 	base := t.TempDir()
 	stg, err := storage.NewLocalStorage(base)
 	require.NoError(t, err)
-	name, data, err := restore.MarshalLogRestoreTableIDsBlocklistFile(0xFFFFFCDEFFFFF, 0xFFFFFFABCFFFF, []int64{1, 2, 3})
+	name, data, err := restore.MarshalLogRestoreTableIDsBlocklistFile(0xFFFFFCDEFFFFF, 0xFFFFFFABCFFFF, 0xFFFFFCCCFFFFF, []int64{1, 2, 3}, []int64{4})
 	require.NoError(t, err)
 	restoreCommitTs, snapshotBackupTs, parsed := restore.ParseLogRestoreTableIDsBlocklistFileName(name)
 	require.True(t, parsed)
@@ -167,18 +167,20 @@ func TestLogRestoreTableIDsBlocklistFile(t *testing.T) {
 	require.NoError(t, err)
 	data, err = stg.ReadFile(ctx, name)
 	require.NoError(t, err)
-	restoreCommitTs, snapshotBackupTs, tableIds, err := restore.UnmarshalLogRestoreTableIDsBlocklistFile(data)
+	blocklist, err := restore.UnmarshalLogRestoreTableIDsBlocklistFile(data)
 	require.NoError(t, err)
-	require.Equal(t, uint64(0xFFFFFCDEFFFFF), restoreCommitTs)
-	require.Equal(t, uint64(0xFFFFFFABCFFFF), snapshotBackupTs)
-	require.Equal(t, []int64{1, 2, 3}, tableIds)
+	require.Equal(t, uint64(0xFFFFFCDEFFFFF), blocklist.RestoreCommitTs)
+	require.Equal(t, uint64(0xFFFFFFABCFFFF), blocklist.SnapshotBackupTs)
+	require.Equal(t, uint64(0xFFFFFCCCFFFFF), blocklist.RewriteTs)
+	require.Equal(t, []int64{1, 2, 3}, blocklist.TableIds)
+	require.Equal(t, []int64{4}, blocklist.DbIds)
 }
 
 func writeBlocklistFile(
 	ctx context.Context, t *testing.T, s storage.ExternalStorage,
-	restoreCommitTs, snapshotBackupTs uint64, tableIds []int64,
+	restoreCommitTs, snapshotBackupTs, rewriteTs uint64, tableIds, dbIds []int64,
 ) {
-	name, data, err := restore.MarshalLogRestoreTableIDsBlocklistFile(restoreCommitTs, snapshotBackupTs, tableIds)
+	name, data, err := restore.MarshalLogRestoreTableIDsBlocklistFile(restoreCommitTs, snapshotBackupTs, rewriteTs, tableIds, dbIds)
 	require.NoError(t, err)
 	err = s.WriteFile(ctx, name, data)
 	require.NoError(t, err)
@@ -197,11 +199,14 @@ func TestCheckTableTrackerContainsTableIDsFromBlocklistFiles(t *testing.T) {
 	base := t.TempDir()
 	stg, err := storage.NewLocalStorage(base)
 	require.NoError(t, err)
-	writeBlocklistFile(ctx, t, stg, 100, 10, []int64{100, 101, 102})
-	writeBlocklistFile(ctx, t, stg, 200, 20, []int64{200, 201, 202})
-	writeBlocklistFile(ctx, t, stg, 300, 30, []int64{300, 301, 302})
-	tableNameByTableID := func(tableID int64) string {
-		return fmt.Sprintf("table_%d", tableID)
+	writeBlocklistFile(ctx, t, stg, 100, 10, 50, []int64{100, 101, 102}, []int64{103})
+	writeBlocklistFile(ctx, t, stg, 200, 20, 60, []int64{200, 201, 202}, []int64{203})
+	writeBlocklistFile(ctx, t, stg, 300, 30, 70, []int64{300, 301, 302}, []int64{303})
+	tableNameByTableId := func(tableId int64) string {
+		return fmt.Sprintf("table_%d", tableId)
+	}
+	dbNameByDbId := func(dbId int64) string {
+		return fmt.Sprintf("db_%d", dbId)
 	}
 	checkTableIDLost := func(tableId int64) bool {
 		return false
@@ -209,26 +214,30 @@ func TestCheckTableTrackerContainsTableIDsFromBlocklistFiles(t *testing.T) {
 	checkTableIDLost2 := func(tableId int64) bool {
 		return true
 	}
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{300, 301, 302}), 250, 300, tableNameByTableID, checkTableIDLost)
+	rewriteTss := make([]uint64, 0)
+	cleanErr := func(rewriteTs uint64) {
+		rewriteTss = append(rewriteTss, rewriteTs)
+	}
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{300, 301, 302}), 250, 300, tableNameByTableId, dbNameByDbId, checkTableIDLost, checkTableIDLost, cleanErr)
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "table_300")
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{200, 201, 202}), 250, 300, tableNameByTableID, checkTableIDLost)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{200, 201, 202}), 250, 300, tableNameByTableId, dbNameByDbId, checkTableIDLost, checkTableIDLost, cleanErr)
 	require.NoError(t, err)
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{200, 201, 202}), 250, 300, tableNameByTableID, checkTableIDLost2)
-	require.Error(t, err)
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{100, 101, 102}), 250, 300, tableNameByTableID, checkTableIDLost)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{200, 201, 202}), 250, 300, tableNameByTableId, dbNameByDbId, checkTableIDLost2, checkTableIDLost2, cleanErr)
 	require.NoError(t, err)
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{100, 101, 102}), 250, 300, tableNameByTableID, checkTableIDLost2)
-	require.Error(t, err)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{100, 101, 102}), 250, 300, tableNameByTableId, dbNameByDbId, checkTableIDLost, checkTableIDLost, cleanErr)
+	require.NoError(t, err)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{100, 101, 102}), 250, 300, tableNameByTableId, dbNameByDbId, checkTableIDLost2, checkTableIDLost2, cleanErr)
+	require.NoError(t, err)
 
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{300, 301, 302}), 1, 25, tableNameByTableID, checkTableIDLost)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{300, 301, 302}), 1, 25, tableNameByTableId, dbNameByDbId, checkTableIDLost, checkTableIDLost, cleanErr)
 	require.NoError(t, err)
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{300, 301, 302}), 1, 25, tableNameByTableID, checkTableIDLost2)
-	require.Error(t, err)
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{200, 201, 202}), 1, 25, tableNameByTableID, checkTableIDLost)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{300, 301, 302}), 1, 25, tableNameByTableId, dbNameByDbId, checkTableIDLost2, checkTableIDLost2, cleanErr)
+	require.NoError(t, err)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{200, 201, 202}), 1, 25, tableNameByTableId, dbNameByDbId, checkTableIDLost, checkTableIDLost, cleanErr)
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "table_200")
-	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{100, 101, 102}), 1, 25, tableNameByTableID, checkTableIDLost)
+	err = restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(ctx, stg, fakeTrackerID([]int64{100, 101, 102}), 1, 25, tableNameByTableId, dbNameByDbId, checkTableIDLost, checkTableIDLost, cleanErr)
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "table_100")
 }
@@ -247,9 +256,9 @@ func TestTruncateLogRestoreTableIDsBlocklistFiles(t *testing.T) {
 	base := t.TempDir()
 	stg, err := storage.NewLocalStorage(base)
 	require.NoError(t, err)
-	writeBlocklistFile(ctx, t, stg, 100, 10, []int64{100, 101, 102})
-	writeBlocklistFile(ctx, t, stg, 200, 20, []int64{200, 201, 202})
-	writeBlocklistFile(ctx, t, stg, 300, 30, []int64{300, 301, 302})
+	writeBlocklistFile(ctx, t, stg, 100, 10, 50, []int64{100, 101, 102}, []int64{103})
+	writeBlocklistFile(ctx, t, stg, 200, 20, 60, []int64{200, 201, 202}, []int64{203})
+	writeBlocklistFile(ctx, t, stg, 300, 30, 70, []int64{300, 301, 302}, []int64{303})
 
 	err = restore.TruncateLogRestoreTableIDsBlocklistFiles(ctx, stg, 50)
 	require.NoError(t, err)

--- a/br/pkg/stream/table_mapping.go
+++ b/br/pkg/stream/table_mapping.go
@@ -104,6 +104,7 @@ func NewTableMappingManager() *TableMappingManager {
 		tempIDCounter:         InitialTempId,
 		tempDefaultKVTableMap: make(map[tableMetaKey]*tableMetaValue),
 		tempDefaultKVDbMap:    make(map[dbMetaKey]*dbMetaValue),
+		noDefaultKVErrorMap:   make(map[uint64]error),
 	}
 }
 

--- a/br/pkg/stream/table_mapping.go
+++ b/br/pkg/stream/table_mapping.go
@@ -93,6 +93,8 @@ type TableMappingManager struct {
 
 	tempDefaultKVTableMap map[tableMetaKey]*tableMetaValue
 	tempDefaultKVDbMap    map[dbMetaKey]*dbMetaValue
+
+	noDefaultKVErrorMap map[uint64]error
 }
 
 func NewTableMappingManager() *TableMappingManager {
@@ -318,9 +320,11 @@ func (tm *TableMappingManager) parseDBValueAndUpdateIdMappingForWriteCf(
 		zap.Uint64("commit-ts", commitTs),
 		zap.String("value", base64.StdEncoding.EncodeToString(value)),
 		zap.Int("temp-default-kv-db-map-size", len(tm.tempDefaultKVDbMap)))
-	return errors.Errorf(errMsgDefaultCFKVLost+"(db id:%d, value %s)",
+	tm.noDefaultKVErrorMap[commitTs] = errors.Errorf(
+		errMsgDefaultCFKVLost+"(db id:%d, value %s)",
 		dbId, base64.StdEncoding.EncodeToString(value),
 	)
+	return nil
 }
 
 func (tm *TableMappingManager) parseDBValueAndUpdateIdMapping(
@@ -482,17 +486,18 @@ func (tm *TableMappingManager) parseTableValueAndUpdateIdMappingForWriteCf(
 		}
 		return tm.parseTableValueAndUpdateIdMapping(dbId, tableId, commitTs, tableValue.info, collector)
 	}
-	log.Error("default cf kv is lost when processing write cf kv for table",
+	log.Warn("default cf kv is lost when processing write cf kv for table",
 		zap.Int64("db-id", dbId),
 		zap.Int64("table-id", tableId),
 		zap.Uint64("start-ts", startTs),
 		zap.Uint64("commit-ts", commitTs),
 		zap.String("value", base64.StdEncoding.EncodeToString(value)),
 		zap.Int("temp-default-kv-table-map-size", len(tm.tempDefaultKVTableMap)))
-	return errors.Errorf(
+	tm.noDefaultKVErrorMap[commitTs] = errors.Errorf(
 		errMsgDefaultCFKVLost+"(db id:%d, table id:%d, value %s)",
 		dbId, tableId, base64.StdEncoding.EncodeToString(value),
 	)
+	return nil
 }
 
 func (tm *TableMappingManager) parseTableValueAndUpdateIdMapping(
@@ -523,6 +528,19 @@ func (tm *TableMappingManager) parseTableValueAndUpdateIdMapping(
 		}
 	}
 	collector.OnTableInfo(dbId, tableId, tableSimpleInfo, commitTs)
+	return nil
+}
+
+func (tm *TableMappingManager) CleanError(rewriteTs uint64) {
+	delete(tm.noDefaultKVErrorMap, rewriteTs)
+}
+
+func (tm *TableMappingManager) ReportIfError() error {
+	for _, err := range tm.noDefaultKVErrorMap {
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
 	return nil
 }
 

--- a/br/pkg/stream/table_mapping_test.go
+++ b/br/pkg/stream/table_mapping_test.go
@@ -1360,3 +1360,13 @@ func TestTableHistoryManagerOutOfOrderTS(t *testing.T) {
 		})
 	}
 }
+
+func TestReportError(t *testing.T) {
+	manager := NewTableMappingManager()
+	manager.noDefaultKVErrorMap[1] = errors.New("test")
+	require.Error(t, manager.ReportIfError())
+	manager.CleanError(2)
+	require.Error(t, manager.ReportIfError())
+	manager.CleanError(1)
+	require.NoError(t, manager.ReportIfError())
+}

--- a/br/pkg/task/restore.go
+++ b/br/pkg/task/restore.go
@@ -270,7 +270,9 @@ type RestoreConfig struct {
 	// [startTs, RestoreTS] is used to `restore log` from StartTS to RestoreTS.
 	StartTS uint64 `json:"start-ts" toml:"start-ts"`
 	// if not specified system will restore to the max TS available
-	RestoreTS       uint64                      `json:"restore-ts" toml:"restore-ts"`
+	RestoreTS uint64 `json:"restore-ts" toml:"restore-ts"`
+	// rewriteTS is the rewritten timestamp of meta kvs.
+	RewriteTS       uint64                      `json:"-" toml:"-"`
 	tiflashRecorder *tiflashrec.TiFlashRecorder `json:"-" toml:"-"`
 	PitrBatchCount  uint32                      `json:"pitr-batch-count" toml:"pitr-batch-count"`
 	PitrBatchSize   uint32                      `json:"pitr-batch-size" toml:"pitr-batch-size"`
@@ -896,7 +898,11 @@ func RunRestore(c context.Context, g glue.Glue, cmdName string, cfg *RestoreConf
 		for tableId := range cfg.PiTRTableTracker.PartitionIds {
 			tableIds = append(tableIds, tableId)
 		}
-		filename, data, err := restore.MarshalLogRestoreTableIDsBlocklistFile(restoreCommitTs, cfg.StartTS, tableIds)
+		dbIds := make([]int64, 0, len(cfg.PiTRTableTracker.DBIds))
+		for dbId := range cfg.PiTRTableTracker.DBIds {
+			dbIds = append(dbIds, dbId)
+		}
+		filename, data, err := restore.MarshalLogRestoreTableIDsBlocklistFile(restoreCommitTs, cfg.StartTS, cfg.RewriteTS, tableIds, dbIds)
 		if err != nil {
 			restoreErr = err
 			return
@@ -1268,13 +1274,14 @@ func runSnapshotRestore(c context.Context, mgr *conn.Mgr, g glue.Glue, cmdName s
 
 	// only run when this full restore is part of the PiTR
 	if isPiTR {
+		snapshotDbMap := client.GetDatabaseMap()
 		snapshotTableMap := client.GetTableMap()
 		// adjust tables to restore in the snapshot restore phase since it will later be renamed during
 		// log restore and will fall into or out of the filter range.
 		err = AdjustTablesToRestoreAndCreateTableTracker(
 			cfg.logTableHistoryManager,
 			cfg.RestoreConfig,
-			client.GetDatabaseMap(),
+			snapshotDbMap,
 			snapshotTableMap,
 			client.GetPartitionMap(),
 			tableMap,
@@ -1288,16 +1295,16 @@ func runSnapshotRestore(c context.Context, mgr *conn.Mgr, g glue.Glue, cmdName s
 			zap.Int("tables", len(tableMap)),
 			zap.Int("db", len(dbMap)))
 
-		tableNameByTableID := func(tableID int64) string {
+		tableNameByTableID := func(tableId int64) string {
 			dbName, tableName, dbID := "", "", int64(0)
 			history := cfg.logTableHistoryManager.GetTableHistory()
-			if locations, exists := history[tableID]; exists {
+			if locations, exists := history[tableId]; exists {
 				if name, exists := cfg.logTableHistoryManager.GetDBNameByID(locations[1].DbID); exists {
 					dbName = name
 				}
 				dbID = locations[1].DbID
 				tableName = locations[1].TableName
-			} else if tableMeta, exists := tableMap[tableID]; exists && tableMeta != nil && tableMeta.Info != nil {
+			} else if tableMeta, exists := tableMap[tableId]; exists && tableMeta != nil && tableMeta.Info != nil {
 				if tableMeta.DB != nil && len(dbName) == 0 {
 					dbName = tableMeta.DB.Name.O
 				}
@@ -1310,7 +1317,16 @@ func runSnapshotRestore(c context.Context, mgr *conn.Mgr, g glue.Glue, cmdName s
 			}
 			return fmt.Sprintf("%s.%s", dbName, tableName)
 		}
-		checkTableIDLost := func(tableId int64) bool {
+		dbNameByDbId := func(dbId int64) string {
+			if dbName, ok := cfg.logTableHistoryManager.GetDBNameByID(dbId); ok {
+				return dbName
+			}
+			if dbInfo, ok := dbMap[dbId]; ok && dbInfo != nil && dbInfo.Info != nil {
+				return dbInfo.Info.Name.O
+			}
+			return ""
+		}
+		checkTableIdLost := func(tableId int64) bool {
 			// check whether exists in log backup
 			if _, exists := cfg.logTableHistoryManager.GetTableHistory()[tableId]; exists {
 				return false
@@ -1321,10 +1337,24 @@ func runSnapshotRestore(c context.Context, mgr *conn.Mgr, g glue.Glue, cmdName s
 			}
 			return true
 		}
+		checkDbIdLost := func(dbId int64) bool {
+			// check whether exists in log backup
+			if _, exists := cfg.logTableHistoryManager.GetDBNameByID(dbId); exists {
+				return false
+			}
+			// check whether exists in snapshot backup
+			if _, exists := snapshotDbMap[dbId]; exists {
+				return false
+			}
+			return true
+		}
 		if err := restore.CheckTableTrackerContainsTableIDsFromBlocklistFiles(
 			ctx, cfg.logRestoreStorage, cfg.PiTRTableTracker, backupMeta.GetEndVersion(), cfg.piTRTaskInfo.RestoreTS,
-			tableNameByTableID, checkTableIDLost,
+			tableNameByTableID, dbNameByDbId, checkTableIdLost, checkDbIdLost, cfg.tableMappingManager.CleanError,
 		); err != nil {
+			return errors.Trace(err)
+		}
+		if err := cfg.tableMappingManager.ReportIfError(); err != nil {
 			return errors.Trace(err)
 		}
 	}

--- a/br/pkg/task/stream.go
+++ b/br/pkg/task/stream.go
@@ -1525,6 +1525,7 @@ func restoreStream(
 	if err != nil {
 		return errors.Trace(err)
 	}
+	cfg.RewriteTS = currentTS
 
 	if err := client.SetCurrentTS(currentTS); err != nil {
 		return errors.Trace(err)

